fsutils/mkfatfs/mkfatfs.c

```diff
@@ -337,7 +338,11 @@ int mkfatfs(FAR const char *pathname, FAR struct fat_format_s *fmt)
 
   /* Allocate a buffer that will be working sector memory */
 
+#ifdef CONFIG_FAT_DMAMEMORY
+  var.fv_sect = (FAR uint8_t *)fat_dma_alloc(var.fv_sectorsize);
+#else
```

Comment:
Sorry this cannot be merged  Applications ares not permitted to call internal to call internal operating system interfaces.  That violates the porttable POSIX interface and will break all PROTECTED and KERNAL biulds that use mkfatfs.

Reply:
Ah, this is why you removed it while moving to apps/ then.

But fat_dma_alloc is defined in https://github.com/apache/incubator-nuttx/blob/master/include/nuttx/fs/fat.h#L115
And there are many other apps that use something like #include <nuttx/*.h>

Anyway what do you suggest ? Is there a fix or I just remove the PR ?

Reply:
@povauboin the better fix is at SPI/DMA driver since the limitation origins from there. SPI/DMA driver should check whether the address is outside it's capability and if so it should utilize(alloc and copy) the special memory internally.
The design of fat_dma_alloc/fat_dma_free is totally wrong:
1.How to handle other file system(e.g. NXFFS, SmartFS, LittleFS...)
2.How to handle the block level access
We should remove them even inside the kernel.

Reply:
> 
> 
> @povauboin the better fix is at SPI/DMA driver since the limitation origins from there. SPI/DMA driver should check whether the address is outside it's capability and if so it can utilize(alloc and copy) the special memory internally.
> The design of fat_dma_alloc/fat_dma_free is totally wrong:
> 1.How to handle other file system(e.g. NXFFS, SmartFS, LittleFS...)
> 2.How to handle the block level access
> We should remove them even inside the kernel.

The solution is already implemented inside the OS for the case of FAT with CONFIG_FAT_DMAMEMORY.  This PR is unnecessary since nothing is broken; mkfatfs works just fine without DAM.  With CONFIG_FAT_DMAMEMORY, DMA will not be used and the filesystem will simply not use it.

In the PX4 case, they do not use SPI for FAT, but an MCI card driver (SD/MMC).  It already implements the DMA.

I agree that the design is wrong.  Different peripherals, however, have different DMA requirements, however.  For example, LPC17 requires some peripheral DMAs to be in one memory bank and other peripheral DMAs to be done in other memory banks.  Some peripherals have different alignment requirements.  Some DMA memory must be allocated from different memory pools.

Block level access still goes through the same peripheral drivers.  I am not sure if it is possible to define a common DMA interface at the block driver level.  But some peripheral interfaces define memory allocation and free methods that could be used at the block driver level for custom peripheral DMA memory allocations.

Reply:
The simplest approach is using bouce buffer I suggest before:
https://lwn.net/Articles/91870/
but we lose some performance due to memory copy(still better than interrupt driven). Another complex approach is exposing dma buffer allocation as Greg's suggestion to userspace or block driver.